pkg/sources/source_manager.go

```diff
@@ -0,0 +1,168 @@
+	handles     map[handle]sourceTemplate
+	handlesLock sync.Mutex
+	// Counters for assigning handle and job IDs for headless managers.
+	headlessHandleCounter int64
+	headlessJobCounter    int64
+	// Pool limiting the amount of concurrent sources running.
+	pool errgroup.Group
+	// Downstream chunks channel to be scanned.
+	outputChunks chan *Chunk
+}
+
+// sourceTemplate is an aggregate struct to hold the Source and its name.
+type sourceTemplate struct {
+	name   string
+	source Source
+}
+
+// apiClient is an interface for optionally communicating with an external API.
+type apiClient interface {
+	// RegisterSource lets the API know we have a source and it returns a unique source ID for it.
+	RegisterSource(ctx context.Context, name string, kind sourcespb.SourceType) (int64, error)
+	// GetJobID queries the API for an existing job or new job ID.
+	GetJobID(ctx context.Context, id int64) (int64, error)
+}
+
+// WithAPI adds an API client to the manager for tracking jobs and progress.
+func WithAPI(api apiClient) func(*SourceManager) {
+	return func(man *SourceManager) { man.api = api }
+}
+
+// WithConcurrency limits the concurrent number of sources a manager can run.
+func WithConcurrency(concurrency int) func(*SourceManager) {
+	return func(man *SourceManager) { man.pool.SetLimit(concurrency) }
+}
+
+// NewManager creates a new manager with the provided options.
+func NewManager(outputChunks chan *Chunk, opts ...func(*SourceManager)) *SourceManager {
+	man := SourceManager{
+		handles:      make(map[handle]sourceTemplate),
+		outputChunks: outputChunks,
+	}
+	for _, opt := range opts {
+		opt(&man)
+	}
+	return &man
+}
+
+// Manage informs the SourceManager to track and manage a Source.
+func (s *SourceManager) Manage(ctx context.Context, name string, source Source) (handle, error) {
+	id, err := s.registerSource(ctx, name, source)
+	if err != nil {
+		return 0, err
+	}
+	s.handlesLock.Lock()
+	defer s.handlesLock.Unlock()
+	if _, ok := s.handles[id]; ok {
+		// TODO: smartly handle this?
+		return 0, fmt.Errorf("generated handle ID '%d' already in use locally", id)
+	}
+	s.handles[id] = sourceTemplate{name, source}
+	return 0, nil
+}
+
+// Run blocks until a resource is available to run the source, then synchronously runs it.
+func (s *SourceManager) Run(ctx context.Context, handle handle, conn *anypb.Any) error {
+	template, ok := s.getTemplate(handle)
+	if !ok {
+		return fmt.Errorf("unrecognized handle")
+	}
+	ch := make(chan error)
+	s.pool.Go(func() error {
+		defer common.Recover(ctx)
+		// TODO: The manager should record these errors.
+		ch <- s.run(ctx, handle, template, conn)
+		return nil
+	})
+	return <-ch
+}
+
+// ScheduleRun blocks until a resource is available to run the source, then
+// asynchronously runs it. Error information is lost in this case.
+func (s *SourceManager) ScheduleRun(ctx context.Context, handle handle, conn *anypb.Any) error {
+	if err := ctx.Err(); err != nil {
+		return err
```

Comment:
Why isn't this check done in `Run`?

Reply:
Since `Run` is synchronous, the context cancellation should be caught downstream and returned as an error. Wouldn't hurt to add it there too though.